dt_core.c

```
@@ -94,6 +96,83 @@ dt_from_ywd(int y, int w, int d) {
     return dt;
 }
 
+#ifdef DT_PARSE_ISO_TNT
+/*
+ * In addition to standard date/month validations we need to check
+ * validity of a day range, so we have a chance to intercept
+ * 32-bit dt_t overflow/underflow. So we merge `dt_valid_*` and `dt_from_*`
+ * set of functions.
+ */
+
+typedef int64_t dt64_t;
+
+bool
+dt_from_yd_checked(int y, int d, dt_t *val) {
+    if (d < 1 || d > dt_days_in_year(y))
+        return false;
+    int64_t y64 = y - 1;
+    int64_t d64 = d;
+    if (y64 < 0) {
+        const int64_t n400 = 1 - y64/400;
+        y64 += n400 * 400;
+        d64 -= n400 * 146097;
+    }
+    dt64_t v = 365 * y64 + y64/4 - y64/100 + y64/400 + d64 + DT_EPOCH_OFFSET;
+    if (v < INT32_MIN || v > INT32_MAX)
+        return false;
+    *val = (dt_t)v;
+    return true;
+}
+
+bool
+dt_from_ymd_checked(int y, int m, int d, dt_t *val) {
+    if (m < 1 || m > 12)
+        return false;
+    if (d < 1 || d > dt_days_in_month(y, m))
+        return false;
+
+    y += m / 12;
+    m %= 12;
+    if (m < 1)
+        y--, m += 12;
+    return dt_from_yd_checked(y, days_preceding_month[LEAP_YEAR(y)][m] + d, val);
+}
+
+bool
+dt_from_yqd_checked(int y, int q, int d, dt_t *val) {
+    if (q < 1 || q > 4)
+        return false;
+    if (d < 1 || d > dt_days_in_quarter(y, q))
+        return false;
+    y += q / 4;
+    q %= 4;
+    if (q < 1)
+        y--, q += 4;
+    return dt_from_yd_checked(y, days_preceding_quarter[LEAP_YEAR(y)][q] + d, val);
+}
+
+bool
+dt_from_ywd_checked(int y, int w, int d, dt_t *val) {
+    if (d < 1 || d > 7)
+        return false;
+    if (w < 1 || w > dt_weeks_in_year(y))
+        return false;
+
+    /* The first week of the year always contains 4 January. */
+    if (!dt_from_yd_checked(y, 4, val))
```

I missed that last time I reviewed this, but looks like dt_from_ywd_checked() won't work for the smallest year:
Jan 4 of the smallest year will always be out of bounds, even when user wants W52.

---

I suggest you introduce `dt64_from_yd`, which returns `dt64_t` for this case, and doesn't check boundaries.

---

That's good you mentioned this boundary check for smallest (and largest) year value. I've realized it's not got parsed properly yet:
```
tarantool> T = date.parse '-5879610W521'
---
- error: 'builtin/datetime.lua:917: could not parse ''-5879610W521'''
...

tarantool> T = date.parse '-5879611W521'
---
- error: 'builtin/datetime.lua:917: could not parse ''-5879611W521'''
...
```

Going to add test both to `c-dt` and `tarantool`. 

P.S.
And I'd rather not copy paste code to yet another `dt_from_yd*` implementation, but rather pass extra argument to `dt_from_yd__checked` which would signal whether we want to raise an error or silently ignore checks for `INT32_MIN`/`INT32_MAX` values.

---

You might extract `dt64_from_yd` from `dt_from_yd_checked` (the one you have now). And simply reuse `dt64_from_yd` in `dt_from_yd_checked`. Anyway, up to you.

---

Though now, having spent some time in debugger, I realize why we needed 64-bit version here (it's under/overflowed if pass `dt_t` for corner cases without boundary checks)

---

Here is final diff with accompanying test for boundary values with W or Q literals:
```diff
diff --git a/dt_core.c b/dt_core.c
index 8cee0c0..e0a75c6 100644
--- a/dt_core.c
+++ b/dt_core.c
@@ -106,8 +106,8 @@ dt_from_ywd(int y, int w, int d) {
 
 typedef int64_t dt64_t;
 
-bool
-dt_from_yd_checked(int y, int d, dt_t *val) {
+static inline bool
+dt64_from_yd_checked(int y, int d, dt64_t *v, bool check) {
     if (d < 1 || d > dt_days_in_year(y))
         return false;
     int64_t y64 = y - 1;
@@ -117,9 +117,21 @@ dt_from_yd_checked(int y, int d, dt_t *val) {
         y64 += n400 * 400;
         d64 -= n400 * 146097;
     }
-    dt64_t v = 365 * y64 + y64/4 - y64/100 + y64/400 + d64 + DT_EPOCH_OFFSET;
-    if (v < INT32_MIN || v > INT32_MAX)
+    *v = 365 * y64 + y64/4 - y64/100 + y64/400 + d64 + DT_EPOCH_OFFSET;
+    if (check && (*v < INT32_MIN || *v > INT32_MAX))
         return false;
+    return true;
+}
+
+bool
+dt_from_yd_checked(int y, int d, dt_t *val) {
+    dt64_t v = 0;
+    if (!dt64_from_yd_checked(y, d, &v, true))
+           return false;
+#ifndef NDEBUG
+    int32_t highw = v >> 32;
+    assert(highw == 0 || highw == -1);
+#endif
     *val = (dt_t)v;
     return true;
 }
@@ -159,10 +171,10 @@ dt_from_ywd_checked(int y, int w, int d, dt_t *val) {
         return false;
 
     /* The first week of the year always contains 4 January. */
-    if (!dt_from_yd_checked(y, 4, val))
+    dt64_t dt = 0;
+    if (!dt64_from_yd_checked(y, 4, &dt, false))
         return false;
 
-    dt64_t dt = *val;
     dt -= dt_dow(dt);
     dt += w * 7 + d - 7;
     if (dt < INT32_MIN || dt > INT32_MAX)
diff --git a/dt_parse_iso.c b/dt_parse_iso.c
index 1711955..9a87882 100644
--- a/dt_parse_iso.c
+++ b/dt_parse_iso.c
@@ -434,20 +434,20 @@ dt_parse_iso_zone_lenient(const char *str, size_t len, int *op) {
 
 #ifdef DT_PARSE_ISO_TNT
 /*
- * Count number of delimiting dashes in date string like
+ * Count number of delimiting dashes, Ws or Qs in date string like
  *  5879611-07-11, or 2012-Q4-85, or 10000W521
  * Allows both ISO8601 and extended Tarantool datetime formats
  */
 static size_t
-count_dashes(const unsigned char *p, size_t i, size_t len) {
+count_delims(const unsigned char *p, size_t i, size_t len) {
     size_t n = 0;
 
     for (; i < len; i++) {
         switch (p[i]) {
         case '0': case '1': case '2': case '3': case '4':
         case '5': case '6': case '7': case '8': case '9':
-        case 'Q': case 'W':
             continue;
+        case 'Q': case 'W':
         case '-':
             n++;
             continue;
@@ -486,30 +486,30 @@ dt_parse_iso_date(const char *str, size_t len, dt_t *dtp) {
         p++;
         len--;
     }
-    dashes_n = count_dashes(p, 0, len);
+    dashes_n = count_delims(p, 0, len);
 #endif
 
     n = count_digits(p, 0, len);
     switch (n) {
 #ifdef DT_PARSE_ISO_TNT
-        case 3: /* -001-01-01 (extended Tarantool range) */
-            if (dashes_n != 2)
+        case 3: /* -001-01-01 | 100W521 (extended Tarantool range) */
+            if (!dashes_n)
                 return 0;
             head_n = 3;
             y = parse_number(p, 0, 3);
             break;
-        case 4: /* -2001 (extended Tarantool range) | 2001-01-01 */
+        case 4: /* -2001 (extended Tarantool range) | 2001-01-01 | 1000W521 */
             y = parse_number(p, 0, 4);
             break;
-        case 5: /* 10000-01-01 (extended Tarantool range) */
-        case 6: /* 109000-01-01 (extended Tarantool range) */
-            if (dashes_n != 2)
+        case 5: /* 10000-01-01 | 10000W521 (extended Tarantool range) */
+        case 6: /* 109000-01-01 | 109000W521 (extended Tarantool range) */
+            if (!dashes_n)
                 return 0;
             y = parse_number(p, 0, n);
             head_n = n;
             break;
-        case 7: /* 5879611-07-11 (extended Tarantool range) */
-            if (dashes_n == 2) {
+        case 7: /* 5879611-07-11 | 1000000W521 (extended Tarantool range) */
+            if (dashes_n > 0) {
                 y = parse_number(p, 0, 7);
                 head_n = n;
                 break;
diff --git a/t/parse_iso_date.c b/t/parse_iso_date.c
index dcf8364..775c408 100644
--- a/t/parse_iso_date.c
+++ b/t/parse_iso_date.c
@@ -35,8 +35,10 @@ const struct good_t {
     {-1000,12, 31, "-1000-12-31",               11 },
     {-10000,12,31, "-10000-12-31",              12 },
     {-5879610,6,22,"-5879610-06-22",            14 },
+    {-5879610,12,27,"-5879610W521",             12 },
     {10000, 1,  1, "10000-01-01",               11 },
     {5879611,7, 1, "5879611-07-01",             13 },
+    {5879611,1, 1, "5879611Q101",               11 },
 #endif
 };
 
@@ -84,7 +86,9 @@ const struct bad_t {
 #endif
     /* fail both for standard and TNT parsing mode */
     {"-5879611-01-01"},  /* Year less than 5879610-06-22 */
+    {"-5879611W011" },  /* Year less than 5879610-06-22 */
     {"5879612-01-01"},  /* Year greater than 5879611-07-11 */
+    {"5879612W521"  },  /* Year greater than 5879611-07-11 */
 };
 
 int 
```